how-to/integrate-with-sso/client/src/auth.ts

```diff
@@ -0,0 +1,263 @@
+import * as auth0 from "auth0-js";
+import { AuthSettings } from "./shapes";
+
+const STORAGE_REALM = "integrate-with-sso";
+const STORE_ACCESS_TOKEN = "token";
+const STORE_AUTH_STATE = "state";
+
+let authSettings: AuthSettings;
+let authWin: OpenFin.Window;
+let authenticatedCallback: (isAuthenticated) => Promise<void>
+let busyCallback: (isBusy) => Promise<void>
+let informationCallback: (info) => void
+
+export async function init(
+  settings: AuthSettings,
+  authenticatedCb: (isAuthenticated) => Promise<void>,
+  busyCb: (isBusy) => Promise<void>,
+  informationCb: (info) => void
+) {
+  authenticatedCallback = authenticatedCb;
+  busyCallback = busyCb;
+  informationCallback = informationCb;
+
+  informationCallback("Initialising the authentication");
+
+  authSettings = settings;
+  if (!authSettings) {
+    informationCallback("Error: Settings missing cannot continue");
+    return;
+  }
+
+  informationCallback("Checking for existing token");
+  const accessToken = loadProperty(STORE_ACCESS_TOKEN);
+  if (!accessToken) {
+    informationCallback("Access token does not exist, show login page");
+    await login();
+  } else {
+    const isValid = await checkTokenValidity(accessToken);
+    if (!isValid) {
+      informationCallback("Access token not valid, show login page");
+      await login();
+    } else {
+      informationCallback("Access token valid, show application");
+      await authenticatedCallback(true);
+    }
+  }
+}
+
+function createWebAuth(): { webAuth: auth0.WebAuth, state: string } {
+  let state = loadProperty(STORE_AUTH_STATE);
+  if (!loadProperty(STORE_AUTH_STATE)) {
+    state = crypto.randomUUID();
+    saveProperty(STORE_AUTH_STATE, state);
+  }
+
+  return {
+    webAuth: new auth0.WebAuth({
+      domain: authSettings.domain,
+      clientID: authSettings.clientId,
+      responseType: "token"
+    }),
+    state
+  };
+}
+
+function loadProperty(propName: string): string {
+  return window.localStorage.getItem(`${STORAGE_REALM}/${propName}`);
+}
+
+function saveProperty(propName: string, value: string): void {
+  window.localStorage.setItem(`${STORAGE_REALM}/${propName}`, value);
```

Comment:
Should we be saving things to local storage?

Reply:
we do

Reply:
https://auth0.com/docs/secure/security-guidance/data-security/token-storage#browser-in-memory-scenarios i think this is blessed